Framework/DataObjects/src/TimeSplitter.cpp

```diff
@@ -339,68 +459,59 @@ std::set<int> TimeSplitter::outputWorkspaceIndices() const {
+ * @return : TimeROI for the input target
  */
-TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
-  // convert indexes less than NO_TARGET to NO_TARGET
-  const int effectiveIndex = std::max<int>(workspaceIndex, NO_TARGET);
+const TimeROI &TimeSplitter::getTimeROI(const int workspaceIndex) const {
+  std::lock_guard<std::mutex> lock(m_mutex);
 
-  TimeROI output;
-  using map_value_type = std::map<DateAndTime, int>::value_type;
-  auto indexFinder = [effectiveIndex](const map_value_type &value) { return value.second == effectiveIndex; };
-  // find the first place this workspace index exists
-  auto iter = std::find_if(m_roi_map.begin(), m_roi_map.end(), indexFinder);
-  // add the ROI found then loop until we reach the end
-  while (iter != m_roi_map.end()) {
-    // add the ROI
-    const auto startTime = iter->first;
-    iter++;
-    // if the next iterator is the end there is nothing to add
-    if (iter != m_roi_map.end()) {
-      const auto stopTime = iter->first;
-      output.addROI(startTime, stopTime);
-    }
-
-    // look for the next place the workspace index occurs
-    iter = std::find_if(iter, m_roi_map.end(), indexFinder);
+  if (!m_validCachedPartialTimeROIs) {
+    rebuildCachedPartialTimeROIs();
   }
+  // convert indexes less than NO_TARGET to NO_TARGET
+  const int effectiveIndex = std::max<int>(workspaceIndex, NO_TARGET);
 
-  // error check that something is there
-  // ignore index being empty is ok
-  if ((workspaceIndex >= 0) && (output.useAll())) {
-    std::stringstream msg;
-    msg << "No regions exist for workspace index " << workspaceIndex;
-  }
+  if (m_cachedPartialTimeROIs.count(effectiveIndex) == 0)
+    return TimeROI::USE_ALL;
 
-  return output;
+  return m_cachedPartialTimeROIs[effectiveIndex];
 }
 
 /**
- * Cast to a vector of SplittingInterval objects
+ * Returns a vector of splitting intervals corresponding to the m_roi_map.
+ * For efficiency, the actual vector calculation should be done only on demand ("lazy") and only when the current vector
+ * is invalid.
+ * @param includeNoTarget : if true/false, the calculated vector will/will not include intervals with NO_TARGET.
+ * @return : a reference to the vector of splitting intervals
  */
-SplittingIntervalVec TimeSplitter::toSplitters(const bool includeNoTarget) const {
-  std::vector<SplittingInterval> output;
-  if (this->empty())
-    return output;
-  auto startIt = m_roi_map.begin();
-  while (std::next(startIt) != m_roi_map.end()) {
-    /// invoke constructor SplittingInterval(DateAndTime &start, DateAndTime &stop, int index)
-    if (includeNoTarget || startIt->second != NO_TARGET)
-      output.push_back({startIt->first, std::next(startIt)->first, startIt->second});
-    std::advance(startIt, 1);
+const Kernel::SplittingIntervalVec &TimeSplitter::getSplittingIntervals(const bool includeNoTarget) const {
+  std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
`TimeSplitter::getSplittingIntervals` should not have a lock. The lock should be within `rebuildCachedSplittingIntervals`.

The reason is that `TimeSplitter::getSplittingIntervals` is [invoked by `TimeSplitter::splitEventVec`](https://github.com/mantidproject/mantid/blob/ewm1943_filter_events_fast_logs/Framework/DataObjects/src/TimeSplitter.cpp#L605), this invoked by the public  `TimeSplitter::splitEventVec`, this invoked by `TimeSplitter::splitEventList`, this invoked  [_in parallel_ for every detector pixel in FilterEvents](https://github.com/mantidproject/mantid/blob/ewm1943_filter_events_fast_logs/Framework/Algorithms/src/FilterEvents.cpp#L1007). If we put a lock, we may as well forget about FilterEvents working in parallel.
Is this conclusion wrong?

Reply:
Good questions. First off, a lock within `rebuildCachedSplittingIntervals` would not achieve the purpose.  In `TimeSplitter::splitEventVec` every thread needs to get the splitting intervals. But `getSplittingIntervals` first asks a question: is the cache valid. Now imagine thousands of threads getting a "no" answer and proceeding to the next step, i.e. `rebuildCachedSplittingIntervals`. While it is true that only one thread, who gets the lock first, will be allowed to rebuild that cache, once that thread is finished, the next thread will rebuild the cache, and then the next one, etc., etc.
The only way to prevent that from happening is to not allow threads check if the cache is valid until the first thread is finished rebuilding the cache.

Reply:
Yes! this is exactly why I later thought that my request to move the lock within was `rebuildCachedSplittingIntervals` wrong.
Thus, we are in agreement and we don't need to meet later on Teams. I'm (yet again) approving this PR